deploy-service/common/src/main/java/com/pinterest/deployservice/handler/GoalAnalyst.java

```diff
@@ -240,8 +240,9 @@ AgentState proposeNewAgentState(PingReportBean report, AgentBean agent) {
         if (agent != null && agent.getState() == AgentState.STOP) {
             // agent has been explicitly STOP, do not override
             if (agent.getDeploy_stage() == DeployStage.STOPPING && FATAL_AGENT_STATUSES.contains(status)) {
```

Comment:
For each of FATAL_AGENT_STATUES - we can test when agent's deploy stage is DeployStage.STOPPING and deployd return one of the FATAL_AGENT_STATUES and verify that behavior is expected.

Looking at original PR, I guess intention was to let it timeout but that's my assumption.

Reply:
Wondering instead of changing deploy_stage to STOPPED here, should this be handled on consumer side.

Reply:
I had the same thought! Trying it [here in PR #889](https://github.com/pinterest/teletraan/pull/889). Long term I do not like this logic in the worker at all, the whole check should be moved into the GoalAnalyst, but this is acceptable for now.